### PR TITLE
Add interpreter for Expm1Op.

### LINKED ISSUE #1102: Add  interpreter for Expm1Op
Following the [testing guidelines](https://github.com/openxla/stablehlo/blob/main/docs/interpreter_checklist.md)

### PR DESCRIPTION
**Pull request by karthikrangasai**
closes #1102 .

We have the following constraints in the spec:

```
(I1) operand is a tensor of floating-point or complex type.
(C1) operand and result have the same type.
```

These constraints are covered by the following tests

```
I1: a) operand is not a tensor of floating-point or complex type. (Covered by ODS).
C1: a) type(operand) != type(result). (Covered by ODS).
```

### CODE REVIEW

File: docs/spec.md [hunk 1]
```diff
@@ -2570,6 +2570,8 @@ produces a `result` tensor. Depending on the element type, does the following:
 // %result: [0.0, 1.71828187]
```
Comment (sdasgup3) [inline docs/spec.md:2570]: update this example in spec.md to replace `f32` to `f64`. Also reflect the same same in the example presented in the description section of StableHLOOps.td file (As the spec.md and ODS should be in sync w.r.t to the example).

File: stablehlo/reference/Element.cpp [hunk 2]
```diff
@@ -728,9 +736,9 @@ Element rem(const Element &e1, const Element &e2) {
         llvm::report_fatal_error("rem(bool, bool) is unsupported");
       },
       [](APFloat lhs, APFloat rhs) {
-        // APFloat::fmod VS APFloat:remainder: the returned value of the latter
-        // is not guaranteed to have the same sign as lhs. So mod() is preferred
-        // here. The returned "APFloat::opStatus" is ignored.
+        // APFloat::fmod VS APFloat:remainder: the returned value of the
```
Comment (ghpvnist) [inline stablehlo/reference/Element.cpp:739]: Was this change intended? The original comment is within the 80-char limit.

File: stablehlo/dialect/StablehloOps.td
Comment (ghpvnist) [inline stablehlo/dialect/StablehloOps.td:1]: Let's also update the example to use the highest representable bit:
```
    Example:
    ```mlir
    %result = stablehlo.exponential_minus_one %operand : tensor<2xf64>
    ```
```

File: stablehlo/dialect/StablehloOps.td [hunk 3]
```diff
@@ -323,7 +323,8 @@ def StableHLO_ExpOp: StableHLO_UnaryElementwiseOp<"exponential",
 }
 
 def StableHLO_Expm1Op: StableHLO_UnaryElementwiseOp<"exponential_minus_one",
-    [Pure, HLO_CompatibleOperandsAndResultType], HLO_FpOrComplexTensor> {
+    [Pure, HLO_CompatibleOperandsAndResultType],  /*exponential_minus_one_c1*/
```
Comment (ghpvnist) [inline stablehlo/dialect/StablehloOps.td:326]:
```
    [Pure, HLO_CompatibleOperandsAndResultType], /*exponential_minus_one_c1*/
	HLO_FpOrComplexTensor /*exponential_minus_one_i1*/> { /*exponential_minus_one_c1*/
```

Feedback:
* typo (double space)
* `StableHLO_UnaryElementwiseOp` class also has the `SameOperandsAndResultShape` which addresses (C1). Added comment following the [checklist](https://github.com/openxla/stablehlo/blob/main/docs/reference_checklist.md).

File: stablehlo/testdata/expm1_shape_float16_20_20.mlir
Comment (ghpvnist) [inline stablehlo/testdata/expm1_shape_float16_20_20.mlir:1]: Revert this change as it fails due to floating-point differences. FYI you can also run all of our CI tooling locally by running the scripts in `build_tools/github_actions/`. In case if you have a powerful machine handy (completely fine if you don't!), for the set of tests in `stablehlo/testdata`, you can run `build_tools/github_actions/ci_build_cmake_llvm.sh <path_to_llvm> <build_dir>` where `<path_to_llvm>` is likely `stablehlo/llvm-project` and `<build_dir>` is `stablehlo/llvm-build`.

File: stablehlo/testdata/expm1_shape_float32_20_20.mlir
Comment (ghpvnist) [inline stablehlo/testdata/expm1_shape_float32_20_20.mlir:1]: Revert this change as it fails due to floating-point differences.

**Review by ghpvnist — CHANGES_REQUESTED** [review]
Similar comment from recent PR, let's write the constraints verbatim from the spec:
From:
```
(I1) operand is a tensor of floating-point or complex type.
(C1) operand and result have the same shape and type.
```
To:
```
(I1) operand is a tensor of floating-point or complex type.
(C1) operand and result have the same type.
```

From:
```
I1: a) operand is not a tensor of floating-point or complex type. (Covered by ODS).
C1: a) shape(operand) != shape(result). (Covered by ODS).
    b) element_type(operand) != element_type(result). (Covered by ODS).
```
To:
```
I1: a) operand is not a tensor of floating-point or complex type. (Covered by ODS).
C1: a) type(operand) != type(result). (Covered by ODS).
```

File: stablehlo/testdata/expm1_shape_float16_20_20.mlir
Comment (karthikrangasai) [reply stablehlo/testdata/expm1_shape_float16_20_20.mlir:1]: Thanks a lot for the command. I will try running it once. If that does not work, then I will have to rely on github actions after creating the PR.

File: stablehlo/reference/Element.cpp (hunk 2 above)
Comment (karthikrangasai) [reply stablehlo/reference/Element.cpp:739]: Apologies, must have been the IDE auto correcting.

File: stablehlo/testdata/expm1_shape_float16_20_20.mlir [hunk 4]
```diff
@@ -1,4 +1,4 @@
-// RUN-DISABLED: stablehlo-translate --deserialize %s.0_9_0.bc | stablehlo-opt -inline | stablehlo-translate --interpret
+// RUN-DISABLED(#1278): stablehlo-translate --deserialize %s.0_9_0.bc | stablehlo-opt -inline | stablehlo-translate --interpret
```
Comment (ghpvnist) [inline stablehlo/testdata/expm1_shape_float16_20_20.mlir:1]: I plan to add this in bulk in #1418. You may revert this file and the file below.

**Review by ghpvnist — CHANGES_REQUESTED** [review]
One minor change and lgtm

**Review by ghpvnist — CHANGES_REQUESTED** [review]

**Review by ghpvnist — APPROVED** [review]

**Review by sdasgup3 — APPROVED** [review]

**Review by ghpvnist — CHANGES_REQUESTED** [review]
Can you rebase this PR? Some changes to look out for: https://github.com/openxla/stablehlo/pull/1440

**Review by ghpvnist — APPROVED** [review]
